File: ros2_project/src/2-1TopicCommunication/two-way_topic_communication_cpp/src/Laowang.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "std_msgs/msg/u_int32.hpp"
#include <functional>

// 创建一个类节点，名字叫做SingleDogNode，继承自Node
class SingleDogNode : public rclcpp::Node
{
public:
    // 构造函数，有一个参数为节点名称
    SingleDogNode(std::string name) : Node(name) // 有参构造函数
    {
        // 打印一句自我介绍
        RCLCPP_INFO(this->get_logger(), "大家好，我是单身狗%s。", name.c_str());

        // 创建一个订阅者来订阅李四的小说，话题名为novel
        sub_novel = this->create_subscription<std_msgs::msg::String>("novel", 10, std::bind(&SingleDogNode::topic_callback, this, std::placeholders::_1));

        // 创建一个发布者来给稿费，话题名为novel_money
        pub_money = this->create_publisher<std_msgs::msg::UInt32>("novel_money", 10);
    }

private:
    // 声明一个订阅者（成员变量），用于订阅小说
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_novel;

    // 声明一个发布者（成员变量），用于订阅完小说后付钱
    rclcpp::Publisher<std_msgs::msg::UInt32>::SharedPtr pub_money;

    // 收到话题数据的回调函数
    void topic_callback(const std_msgs::msg::String::SharedPtr msg)
    {
        // 1. 订阅者接收数据（接收数据后发布消息）
        RCLCPP_INFO(this->get_logger(), "朕已阅：'%s'", msg->data.c_str());

        // 2. 订阅者发布数据（这里是订阅者看完小说，给发布者打钱，其实就是发布者）
        // 逻辑贯通：老王接收到李四的小说后，就应该给他稿费
        std_msgs::msg::UInt32 money;
        money.data = 10;
        pub_money->publish(money);
        RCLCPP_INFO(this->get_logger(), "我已经发送%d元稿费", money.data);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);                                   // 初始化rclcpp
    auto node = std::make_shared<SingleDogNode>("LaoWang");     // 新建一个节点
    rclcpp::spin(node);                                // 保持节点运行，检测时候收到退出指令
    rclcpp::shutdown();                                         // rcl关闭
    return 0;
}

File: ros2_project/src/2-1TopicCommunication/two_way_topic_communication_py/two_way_topic_communication_py/Laowang.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from std_msgs.msg import UInt32

class SingleDogNode(Node):

    # 构造函数
    def __init__(self,name):
        super().__init__(name)

        # 输出日志，让我们知道节点启动了
        self.get_logger().info(f"单身狗节点{name}启动")
        
        # 创建一个订阅者来订阅李四的小说，话题名为novel
        self.sub_novel = self.create_subscription(String,"novel",self.sub_novel_callback,10)

        # 创建一个发布者来给稿费，话题名为novel_money
        self.pub_money = self.create_publisher(UInt32,"novel_money",10)

    # 订阅者(小说)的回调函数
    def sub_novel_callback(self,msg:String):
        # 1. 订阅者接收数据（接收数据后发布消息）
        self.get_logger().info(f"我收到小说：{msg.data}")
        
        # 2. 订阅者发布数据（这里是订阅者看完小说，给发布者打钱，其实就是发布者）
        # 逻辑贯通：老王接收到李四的小说后，就应该给他稿费
        money = UInt32()
        money.data = 10 # 表示10元
        self.pub_money.publish(money)
        self.get_logger().info(f"我已经发送{money.data}元稿费")
    

def main(args=None):
    rclpy.init(args=args)               # 初始化rclpy
    node = SingleDogNode("LaoWang")     # 新建一个节点
    rclpy.spin(node)                    # 保持节点运行，检测是否收到退出指令(Ctrl+C)
    rclpy.shutdown()                    # 当接收到退出指令时，退出spin函数，执行shutdown函数来关闭rclpy

